oidc/clientassertion/client_assertion_test.go

```diff
@@ -0,0 +1,287 @@
+				WithClientSecret("ssshhhh", "test-alg"),
+			},
+			check: func(t *testing.T, ca *ClientAssertion) {
+				require.Equal(t, map[string]string{"h1": "v1", "h2": "v2"}, ca.headers)
+			},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			ca, err := New(tc.cid, tc.aud, tc.opts...)
+
+			require.NoError(t, err)
+			require.NotNil(t, ca)
+			require.Equal(t, tc.cid, ca.clientID)
+			require.Equal(t, tc.aud, ca.audience)
+
+			if tc.check != nil {
+				tc.check(t, ca)
+			}
+
+		})
+	}
+}
+
+func TestValidate(t *testing.T) {
+	tCid := "test-client-id"
+	tAud := []string{"test-audience"}
+	cases := []struct {
+		name string
+		cid  string
+		aud  []string
+		opts []Option
+		errs []error
+	}{
+		{
+			name: "missing everything",
+			errs: []error{ErrMissingClientID, ErrMissingAudience, ErrMissingAlgorithm, ErrMissingKeyOrSecret},
+		},
+		{
+			name: "missing client id",
+			aud:  tAud,
+			errs: []error{ErrMissingClientID},
+			opts: []Option{
+				WithRSAKey(&rsa.PrivateKey{}, "algo"),
+			},
+		},
+		{
+			name: "missing audience",
+			cid:  tCid,
+			errs: []error{ErrMissingAudience},
+			opts: []Option{
+				WithRSAKey(&rsa.PrivateKey{}, "algo"),
+			},
+		},
+		{
+			name: "missing client and secret",
+			cid:  tCid, aud: tAud,
+			errs: []error{ErrMissingAlgorithm, ErrMissingKeyOrSecret},
+		},
+		{
+			name: "both client and secret",
+			cid:  tCid, aud: tAud,
+			opts: []Option{
+				WithRSAKey(&rsa.PrivateKey{}, "algo"),
+				WithClientSecret("ssshhhh", "algo"),
+			},
+			errs: []error{ErrBothKeyAndSecret},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			// New() runs Validate()
+			ca, err := New(tc.cid, tc.aud, tc.opts...)
+
+			require.NotNil(t, err)
+			require.ErrorContains(t, err, "validation error:")
+
+			err = errors.Unwrap(err) // New wraps the error from Validate() with fmt.Errorf("%w")
+			assertJoinedErrs(t, tc.errs, err)
+
+			require.Nil(t, ca)
+
+		})
+	}
+}
+
+func TestSignedToken(t *testing.T) {
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+	pub, ok := key.Public().(*rsa.PublicKey)
+	require.True(t, ok, "couldn't get rsa.PublicKey from PrivateKey")
+
+	cases := []struct {
+		name     string
+		claimKey any // []byte or pubkey; we'll use this to check the signature
+		opts     []Option
+		err      error
+	}{
+		{
+			name:     "valid secret",
+			claimKey: []byte("ssshhhh"),
+			opts: []Option{
+				WithClientSecret("ssshhhh", "HS256"),
+				WithKeyID("test-key-id"),
+				WithHeaders(map[string]string{"xtra": "headies"}),
+			},
+		},
+		{
+			name:     "valid key",
+			claimKey: pub,
+			opts: []Option{
+				WithRSAKey(key, "RS256"),
+				WithKeyID("test-key-id"),
+				WithHeaders(map[string]string{"xtra": "headies"}),
+			},
+		},
+		{
+			name:     "invalid alg",
+			claimKey: pub,
+			opts: []Option{
+				WithRSAKey(key, "ruh-roh"),
+			},
+			err: jose.ErrUnsupportedAlgorithm,
```

Comment:
Do we want to make this error part of our public API? Maybe it should be intercepted and converted to one from this package?

Reply:
Haha that's totally fair - I think if we go down the route of only accepting the HMAC algorithms, we could define that subset as a constant and make that the input to this function instead.